tarantool_test.go

```diff
@@ -2830,6 +2830,330 @@ func TestStream_DoWithClosedConn(t *testing.T) {
+			}
+			if event.Key != key {
+				t.Errorf("Unexpected event key: %s", event.Key)
+			}
+			if event.Value.(string) != val {
+				t.Errorf("Unexpected event value: %v", event.Value)
+			}
+		case <-time.After(time.Second):
+			t.Fatalf("Failed to get watch event %d", i)
+		}
+	}
+}
+
+func TestConnection_NewWatcher_multiOnKey(t *testing.T) {
+	test_helpers.SkipIfWatchersUnsupported(t)
+
+	const key = "TestConnection_NewWatcher_multiOnKey"
+	const value = "bar"
+
+	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	defer conn.Close()
+
+	events := []chan WatchEvent{
+		make(chan WatchEvent),
+		make(chan WatchEvent),
+	}
+	for _, ch := range events {
+		defer close(ch)
+	}
+
+	for _, ch := range events {
+		channel := ch
+		watcher, err := conn.NewWatcher(key, func(event WatchEvent) {
+			channel <- event
+		})
+		if err != nil {
+			t.Fatalf("Failed to create a watch: %s", err)
+		}
+		defer watcher.Unregister()
+	}
+
+	for i, ch := range events {
+		select {
+		case <-ch: // Skip an initial event.
+		case <-time.After(2 * time.Second):
+			t.Fatalf("Failed to skip watch event for %d callback", i)
+		}
+	}
+
+	_, err := conn.Do(NewBroadcastRequest(key).Value(value)).Get()
+	if err != nil {
+		t.Fatalf("Failed to send a broadcast request: %s", err)
+	}
+
+	for i, ch := range events {
+		select {
+		case event := <-ch:
+			if event.Conn != conn {
+				t.Errorf("Unexpected event connection: %v", event.Conn)
+			}
+			if event.Key != key {
+				t.Errorf("Unexpected event key: %s", event.Key)
+			}
+			if event.Value.(string) != value {
+				t.Errorf("Unexpected event value: %v", event.Value)
+			}
+		case <-time.After(2 * time.Second):
+			t.Fatalf("Failed to get watch event from callback %d", i)
+		}
+	}
+}
+
+func TestWatcher_Unregister(t *testing.T) {
+	test_helpers.SkipIfWatchersUnsupported(t)
+
+	const key = "TestWatcher_Unregister"
+	const value = "bar"
+
+	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	defer conn.Close()
+
+	events := make(chan WatchEvent)
+	defer close(events)
+	watcher, err := conn.NewWatcher(key, func(event WatchEvent) {
+		events <- event
+	})
+	if err != nil {
+		t.Fatalf("Failed to create a watch: %s", err)
+	}
+
+	<-events
+	watcher.Unregister()
+
+	_, err = conn.Do(NewBroadcastRequest(key).Value(value)).Get()
+	if err != nil {
+		t.Fatalf("Got broadcast error: %s", err)
+	}
+
+	select {
+	case event := <-events:
+		t.Fatalf("Get unexpected events: %v", event)
+	case <-time.After(time.Second):
+	}
+}
+
+func TestConnection_NewWatcher_concurrent(t *testing.T) {
+	test_helpers.SkipIfWatchersUnsupported(t)
+
+	const testConcurrency = 1000
+	const key = "TestConnection_NewWatcher_concurrent"
+	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	defer conn.Close()
+
+	var wg sync.WaitGroup
+	wg.Add(testConcurrency)
+
+	var ret error
+	for i := 0; i < testConcurrency; i++ {
+		go func() {
+			defer wg.Done()
+
+			watcher, err := conn.NewWatcher(key, func(event WatchEvent) {})
```

Comment:
I'm bad at Go concurrency (I think this idea will reflect in >=50% of my comments in this review), but what could possibly go wrong here?

Reply:
The tests starts 1000 goroutines. They most likely run in parallel more or less.

This test helped me find the deadlock. In fact, a lot of things can go wrong in the case of concurrent code. But yeah, this test doesn't check for anything specific.